cng/aes.go

```diff
@@ -119,34 +119,45 @@ func (c *aesCipher) NewGCMTLS() (cipher.AEAD, error) {
 	return newGCM(c.key, true)
 }
 
-type aesCBC struct {
-	kh      bcrypt.KEY_HANDLE
-	iv      [aesBlockSize]byte
-	encrypt bool
+type cbcCipher struct {
+	kh bcrypt.KEY_HANDLE
+	// us aesBlockSize as the max block size for all ciphers
+	// to avoid allocations.
+	iv        [aesBlockSize]byte
+	blockSize int
+	encrypt   bool
 }
 
-func newCBC(encrypt bool, key, iv []byte) *aesCBC {
-	kh, err := newCipherHandle(bcrypt.AES_ALGORITHM, bcrypt.CHAIN_MODE_CBC, key)
+func newCBC(encrypt bool, alg string, key, iv []byte) *cbcCipher {
+	kh, err := newCipherHandle(alg, bcrypt.CHAIN_MODE_CBC, key)
 	if err != nil {
 		panic(err)
 	}
-	x := &aesCBC{kh: kh, encrypt: encrypt}
+	x := &cbcCipher{kh: kh, encrypt: encrypt}
+	switch alg {
+	case bcrypt.AES_ALGORITHM:
+		x.blockSize = aesBlockSize
+	case bcrypt.DES_ALGORITHM:
+		x.blockSize = desBlockSize
+	default:
+		panic("invalid algorithm: " + alg)
+	}
 	x.SetIV(iv)
-	runtime.SetFinalizer(x, (*aesCBC).finalize)
+	runtime.SetFinalizer(x, (*cbcCipher).finalize)
 	return x
 }
 
-func (x *aesCBC) finalize() {
+func (x *cbcCipher) finalize() {
 	bcrypt.DestroyKey(x.kh)
 }
 
-func (x *aesCBC) BlockSize() int { return aesBlockSize }
+func (x *cbcCipher) BlockSize() int { return x.blockSize }
 
-func (x *aesCBC) CryptBlocks(dst, src []byte) {
+func (x *cbcCipher) CryptBlocks(dst, src []byte) {
 	if subtle.InexactOverlap(dst, src) {
 		panic("crypto/cipher: invalid buffer overlap")
 	}
```

Comment:
Upstream's version allows some more buffer reuse here too: https://github.com/golang/go/blob/caf9e15fb76631ca692f4779135aec5b82a68370/src/crypto/cipher/cbc.go#L73-L81

```go
	if alias.InexactOverlap(dst[:len(src)], src) {
		panic("crypto/cipher: invalid buffer overlap")
	}
```

I'm thinking it'd make the most sense to go through all of these at once (in Boring) then apply the result to OpenSSL+CNG after spotting everything, leaving this PR as it is for now.